### PR TITLE
Optimizations for diagonal `HermOrSym`

### PR DESCRIPTION
**Pull request by dkarrasch**
The added checks should generically fail quickly, but the benefits of performance in the diagonal case are presumably dramatic. Somewhat related to https://github.com/JuliaLang/LinearAlgebra.jl/issues/922.

### CODE REVIEW

File: stdlib/LinearAlgebra/src/symmetric.jl [hunk 1]
```diff
@@ -578,9 +579,12 @@ end
...
 function dot(x::AbstractVector, A::RealHermSymComplexHerm, y::AbstractVector)
     require_one_based_indexing(x, y)
-    (length(x) == length(y) == size(A, 1)) || throw(DimensionMismatch())
+    n = length(x)
+    (n == length(y) == size(A, 1)) || throw(DimensionMismatch())
     data = A.data
-    r = zero(eltype(x)) * zero(eltype(A)) * zero(eltype(y))
+    r = dot(zero(eltype(x)), zero(eltype(A)), zero(eltype(y)))
+    iszero(n) && return r
+    isdiag(A) && return dot(x, Diagonal(A), y)
```
Comment (stevengj) [inline stdlib/LinearAlgebra/src/symmetric.jl:587]: It doesn't make sense to me to have a generic `isdiag` check in `dot` — isn't it the same complexity as the `dot` itself?

Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/symmetric.jl:587]: You're absolutely right. I should've run benchmarks on the computer, and not just in my head. 🤦  If anything, then we should add what you're suggesting.

**Comment by dkarrasch** [comment]
The factorization change is (not surprisingly) very beneficial, even for small matrices.

File: stdlib/LinearAlgebra/src/symmetric.jl [hunk 2]
```diff
@@ -240,6 +240,7 @@ end
 
 diag(A::Symmetric) = symmetric.(diag(parent(A)), sym_uplo(A.uplo))
 diag(A::Hermitian) = hermitian.(diag(parent(A)), sym_uplo(A.uplo))
+isdiag(A::HermOrSym) = isdiag(parent(A))
```
Comment (stevengj) [inline stdlib/LinearAlgebra/src/symmetric.jl:243]: I'm not sure this is correct.  For example:
```jl
julia> A = Symmetric(tril!(rand(3,3)))
3×3 Symmetric{Float64, Matrix{Float64}}:
 0.321068  0.0       0.0
 0.0       0.268771  0.0
 0.0       0.0       0.355687

julia> isdiag(A)
true

julia> isdiag(parent(A))
false
```

Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/symmetric.jl:243]: Yes, I actually thought about wrapping it in the appropriate triangular type, but then we don't have specialized `isdiag` for that. I'll simply remove this line. Thanks for catching it!

Comment (stevengj) [reply stdlib/LinearAlgebra/src/symmetric.jl]: You could have
```jl
isdiag(::HermOrSym{<:Any,<:Diagonal}) = true
```
I guess

Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/symmetric.jl]: I realized we do have specialized code. For AbstractMatrix, we have the following call chain

```julia
isdiag(A) -> isbanded(A, 0, 0) -> istriu(A, 0) && istril(A, 0)
```

and one of the last two checks is basically for free for `*Triangular` (as expected)!

Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/symmetric.jl]: This is not very important, but still nice to have:

```julia
julia> using LinearAlgebra, BenchmarkTools

julia> A = Symmetric(diagm(0 => ones(1000)));

julia> @btime isdiag($A)
  723.039 μs (1 allocation: 16 bytes) # vs. 2.860 ms (0 allocations: 0 bytes) on v1.8.4
true
```

The single allocation is due to the triangular wrapper.

**Comment by dkarrasch** [comment]
I believe this is in good shape now. @stevengj Can you please take another look?

**Pull request merged**